### PR TITLE
More multi target search fixes

### PR DESCRIPTION
**Pull request by dirkkul**
### What's being changed:

- Concurrently resolve search vectors - they can involve calls to vectorizers (nearText) and should be done concurrently
- Fixes (and tests) for multi node setups
- Only search for IDs concurrently, combine them based on the distances and THEN create objects/filter etc
- Last two commits adds an extra test (no fixes needed)

### Review checklist

- [ ] Documentation has been updated, if necessary. Link to changed documentation:
- [ ] Chaos pipeline run or not necessary. Link to pipeline:
- [ ] All new code is covered by tests where it is reasonable.
- [ ] Performance tests have been run or not necessary.

### CODE REVIEW

File: adapters/clients/remote_index.go [hunk 1]
```diff
@@ -403,10 +405,11 @@ func (c *RemoteIndex) SearchShard(ctx context.Context, host, index, shard string
 	cursor *filters.Cursor,
 	groupBy *searchparams.GroupBy,
 	additional additional.Properties,
+	multiTargetCombination *dto.TargetCombination,
```
Comment (antas-marcin) [inline adapters/clients/remote_index.go:408]: let's use one convention for naming of this variable: `targetCombination`

File: adapters/repos/db/index.go [hunk 2]
```diff
@@ -1609,7 +1609,7 @@ func (i *Index) objectVectorSearch(ctx context.Context, searchVectors [][]float3
 			} else {
 				res, resDists, nodeName, err = i.remote.SearchShard(ctx,
 					shardName, searchVectors, targetVectors, limit, filters,
-					nil, sort, nil, groupBy, additional, i.replicationEnabled())
+					nil, sort, nil, groupBy, additional, i.replicationEnabled(), multiTargetCombination)
```
Comment (antas-marcin) [inline adapters/repos/db/index.go:1612]: would be great to rename it to `targetCombination`

File: adapters/repos/db/shard_combine_multi_target.go [hunk 3]
```diff
@@ -16,55 +16,53 @@ import (
...
 
 type DistanceForVector interface {
-	VectorDistanceForQuery(ctx context.Context, id strfmt.UUID, searchVectors [][]float32, targets []string) ([]float32, error)
+	VectorDistanceForQuery(ctx context.Context, id uint64, searchVectors [][]float32, targets []string) ([]float32, error)
 }
 
-type objAndDistance struct {
-	obj      *storobj.Object
+type idAndDistance struct {
+	obj      uint64
```
Comment (antas-marcin) [inline adapters/repos/db/shard_combine_multi_target.go:35]: can we also change the name from `obj` to `id`?

File: adapters/repos/db/shard_combine_multi_target.go [hunk 4]
```diff
@@ -134,61 +153,59 @@ func CombineMultiTargetResults(ctx context.Context, shard DistanceForVector, log
...
-	combinedResults := make(map[strfmt.UUID]objAndDistance, len(results[0]))
+	combinedResults := make(map[uint64]idAndDistance, len(results[0]))
 	for i := range results {
 		if len(results[i]) > len(dists[i]) {
 			return nil, nil, fmt.Errorf("number of results does not match number of distances")
 		}
-		var localIDs map[strfmt.UUID]*storobj.Object
+		var localIDs map[uint64]struct{}
+
+		if multiTargetCombination.Type != dto.Minimum {
```
Comment (antas-marcin) [inline adapters/repos/db/shard_combine_multi_target.go:185]: would be great to use the same name convention: `targetCombination`

File: adapters/repos/db/shard_combine_multi_target.go (hunk 3 above)
Comment (antas-marcin) [reply adapters/repos/db/shard_combine_multi_target.go:35]: or maybe more appropriate would be `docId`?

File: usecases/sharding/remote_index.go [hunk 5]
```diff
@@ -250,14 +252,15 @@ func (ri *RemoteIndex) SearchShard(ctx context.Context, shard string,
...
+	multiTargetCombination *dto.TargetCombination,
 ) ([]*storobj.Object, []float32, string, error) {
 	type pair struct {
 		first  []*storobj.Object
 		second []float32
 	}
 	f := func(node, host string) (interface{}, error) {
 		objs, scores, err := ri.client.SearchShard(ctx, host, ri.class, shard,
-			queryVec, targetVector, limit, filters, keywordRanking, sort, cursor, groupBy, adds)
+			queryVec, targetVector, limit, filters, keywordRanking, sort, cursor, groupBy, adds, multiTargetCombination)
```
Comment (antas-marcin) [inline usecases/sharding/remote_index.go:263]: would rename to `targetCombination`

File: usecases/traverser/explorer.go [hunk 6]
```diff
@@ -235,18 +237,28 @@ func (e *Explorer) getClassVectorSearch(ctx context.Context,
...
+				searchVectorParam = params.NearVector
+			} else if searchVectorParams != nil {
+				searchVectorParam = searchVectorParams[i]
+			}
 
-		searchVectors[i], err = e.vectorFromParamsForTaget(ctx, searchVectorParam, params.NearObject, params.ModuleParams, params.ClassName, params.Tenant, targetVectors[i])
-		if err != nil {
-			return nil, nil, errors.Errorf("explorer: get class: vectorize search vector: %v", err)
-		}
+			searchVectors[i], err = e.vectorFromParamsForTaget(ctx, searchVectorParam, params.NearObject, params.ModuleParams, params.ClassName, params.Tenant, targetVectors[i])
```
Comment (antas-marcin) [inline usecases/traverser/explorer.go:252]: shouldn't here be mutex used since we change this operation from sequential to parallel?

Comment (antas-marcin) [reply usecases/traverser/explorer.go:252]: or this is a safe operation since we use index?

Comment (dirkkul) [reply usecases/traverser/explorer.go:252]: Not needed - each goroutine only writes into "its" index which is preallocated and there are no reads before the error group finishes

**Review by antas-marcin — APPROVED** [review]

**Pull request merged**